llvm/include/llvm/IR/RepoDefinition.h

```diff
@@ -48,66 +48,76 @@ void set(GlobalObject *GO, DigestType const &D);
 /// GO does not contain the RepoDefinition metadata.
 std::pair<DigestType, bool> get(const GlobalObject *GO);
 
-/// A structure of a global object (GO) information.
+/// A structure of a global object (GO) information, which is used to calculate
+/// the GO's digest.
 struct GOInfo {
   /// The InitialDigest is the hash of an object itself (a function or global
   /// variable) without consideration for any references to other objects.
   DigestType InitialDigest;
-  /// The Contributions of an object (X) are those objects (Y) which
-  /// transitively reference X and where a potential optimisation to either X or
-  /// any of Y may invalidate both.
-  GOVec Contributions;
+
   /// The Dependencies of an object are those objects which it transitively
-  /// references but are not Contributions.
+  /// references or the reverse-contributions, which is converted from the
+  /// Contributions.
   GOVec Dependencies;
 
   GOInfo() = default;
-  GOInfo(DigestType &&Digest, GOVec &&Contributions, GOVec &&Dependencies)
+  GOInfo(DigestType &&Digest, GOVec &&Dependencies)
       : InitialDigest(std::move(Digest)),
-        Contributions(std::move(Contributions)),
         Dependencies(std::move(Dependencies)) {}
 
 #ifndef NDEBUG
   void dump() const {
     dbgs() << "GOInfo:";
     dbgs() << "\n\tInitial Digest: " << InitialDigest.digest();
-    dbgs() << "\n\tContributions: [ ";
+    dbgs() << "\n\tDependencies: [ ";
     bool first = true;
-    for (const auto &Contribution : Contributions) {
+    for (const auto &Dependent : Dependencies) {
       if (!first)
         dbgs() << ",";
-      dbgs() << Contribution->getName();
+      dbgs() << Dependent->getName();
       first = false;
     }
-    dbgs() << "]";
-    dbgs() << "\n\tDependencies: [ ";
-    first = true;
-    for (const auto &Dependent : Dependencies) {
+    dbgs() << "]\n";
+  }
+#endif
+};
+
+/// A structure of a global object (GO) detailed information, which includes the
+/// GO's necessary information (`GOInfo`) for the hash calculation and its
+/// additional `Contributions` information.
+struct GODetailedInfo {
```

Comment:
Can you clarify the relationship between GOInfo and GODetailedInfo? Will a GOInfo within a GODetailedInfo have (reverse-)Contributions in its Dependencies, or will they be in either or.

I feel like a design pattern may fit here (e.g. GODetailedInfo being a decorator of GOInfo), but I don't understand the usage enough yet.

Reply:
The `HashCalculator` is used to calculate the GO’s initial digest, its Contributions and its Dependencies. The result is saved to the `GODetailedInfo` structure. Here, the `Dependencies` only include the partial dependencies which can be found/seen when scanning the GO’s definition.

The `GOInfo` is used to calculate the GO’s final digest. It contains the GO’s initial digest and all/full dependencies. All dependencies contain 1) **the GO’s partial dependencies**, which are given in the `GODetailedInfo` and 2) **the reverse-contributions**, which are converted from the stored `Contributions` in the `GODetailedInfo`.